### ISSUE TITLE
performance is bad

### ISSUE THREAD

**Issue by stevecrozz** [errbit/errbit#885]
Cursory measurements indicate errbit performance is quite bad. We should explore this and find some ways to improve both read and write performance. The first order of business will be to create a test rig where we can gather some metrics about exactly how bad our performance problem is. If we get a good baseline, we can weigh future performance-related decisions against a baseline.

**Comment by arthurnn** [comment]
:+1: :+1: :+1:

**Comment by stevecrozz** [comment]
@arthurnn I have started a little project at https://github.com/stevecrozz/errbit_loader to help measure errbit performance. There's only one test in it now which measures the time it takes to create X notices, Y at a time using the 2.3 airbrake compatible API. Do you have any other ideas for meaningful metrics to gather?

**Comment by arthurnn** [comment]
I think the order of important things performance related are:
1. Notice creation speed
2. Notice page speed
3. Delete/Archive Notice speed.

**Comment by stevecrozz** [comment]
Here's some preliminary results running against a free heroku dyno. The results are really not terrible, but I expect performance to degrade as the database fills up. So I plan to repeat this many times and gather results over time as size of the db increases:
```
CreateNoticesTest
----------------------------------------------------------------------
Concurrency: 8
Count: 100

errors             0
max         1.057916
min         0.554949
avg         0.659651
p90         0.711527
p95         0.800747
p99         1.021317

Total time taken: 8.435283459 seconds
----------------------------------------------------------------------

SearchNoticesTest
----------------------------------------------------------------------
Concurrency: 8
Count: 100

errors             0
max         0.955028
min         0.442574
avg         0.542977
p90         0.679447
p95         0.748714
p99         0.865984

Total time taken: 6.986034781 seconds
----------------------------------------------------------------------
```

**Comment by stevecrozz** [comment]
Here's what I have so far: http://lithostech.com/2015/06/measuring-errbit-performance/

I think this is enough to get started on some actual performance improvements.

**Comment by arthurnn** [comment]
Nice..
I think the next step would be worth trying to use elastic search or pg to compare the results.

**Comment by rud** [comment]
Using a postgresql backend would massively simplify the maintenance of an errbit installation (for me, quite possibly for others). Would be a very interesting simplification direction to take the project in. Does anybody need a hand in moving forward with this?

**Comment by stevecrozz** [comment]
I've got some experiments going on in https://github.com/stevecrozz/errbit/tree/mongoid5. I believe I have identified at least one major performance issue with errbit and the way we're using mongo. I've already reduced the number of mongo queries needed to insert a new notice from nine to five, but I'm willing to look into other options.

@rud: The errbit community has already considered a move to Postgres and that did result in an Errbit fork (see #614). This was a very popular idea, although never merged back to Errbit and its now significantly diverged from Errbit to the point where I believe it would need to be redone from scratch.

If you're willing to work on this, then I'm willing to lend support. Out of curiosity, why do you say this would massively simplify the maintenance of an errbit installation?

**Comment by rud** [comment]
@stevecrozz sad to see issue #614, didn't look like a productive contribution.

The simplification would be me only having to manage PostgreSQL and ruby-processes, and not be responsible for another completely different datastore. I just haven't invested enough time to get to know MongoDB to be entirely comfortable around it the way I am with PostgreSQL.

I could find some time to help out with a migration to a PostgreSQL backend, would probably be a gradual introduction with dual data-stores during the migration process. Would the intended end-goal be to completely remove the support for MongoDB, or would a double backend be a preferred end-state? Personally I'd be in favour of migrating to only support PostgreSQL, once the migration-path was clear.

Thoughts on this?

**Comment by stevecrozz** [comment]
Deployment simplicity is a high priority for me. So whichever way we decide to go, I can't see wanting to end up in a mongo+postgres scenario. But at least for this ticket, I want to maintain focus on performance. I'm tracking down performance issues now and as much as I would like to, I haven't been able to place much blame on the data store.

If I could wave a magic wand and replace Errbit's mongo backend with postgres, I would probably do it because our users were outspoken about this in #614, but I think this effort is a separate one unless we can demonstrate that changing from mongo to postgres will cause a measurable performance improvement for Errbit.

But that's not to minimize the idea of moving to postgres. I do support that idea. @rud if you're willing to put in the time, let's create a new issue to track this effort and you can get started. I do have a big experimental effort open right now with lots of changes related to this ticket, so you may want to wait a couple weeks to see if this effort gets merged and avoid a bunch of conflicts down the road.

**Comment by rud** [comment]
I'm glad we're in agreement about a full migration being desirable after your optimization completes. Holding off starting the effort for now.

**Comment by arthurnn** [comment]
thanks @stevecrozz and @rud for taking care of this.
Would be awesome to grab stats tho, before/after the migration, and prove that PG was indeed N times faster.
I honestly dont care much about the deploy side, as for me to deploy a PG or a MongoDB server has the same difficulty. 
I think we should focus on performance, and specially on insert performance, as that it is 80% use case of errbit.

**Comment by stevecrozz** [comment]
Got a bit more info coming out: http://lithostech.com/2015/07/why-errbit-is-slow/

The summary of my findings so far is that Errbit seems to be using a lot of memory which takes time to both allocate and garbage collect. I'm suspicious of mongoid, but I don't yet have any specific details. Next up is some memory usage profiling.

**Comment by stevecrozz** [comment]
Finally have some results to share: http://lithostech.com/2015/07/inserting-errors-with-the-quickness/

I have an experimental branch of Errbit where I've made a number of improvements that allow for much better throughput when adding errors: https://github.com/stevecrozz/errbit/tree/mongoid5

**Comment by arrowcircle** [comment]
Migration to pg will be awesome

**Comment by rud** [comment]
@stevecrozz: do you have a PR ready for review for merging your work?

**Comment by stevecrozz** [comment]
Yep @rud: I merged my changes this weekend. Everything I had is in master now. The performance improvement is solid. I'll write a wrap-up blog entry this week to highlight the changes. And I guess this ticket is solved as of #917

**Comment by rud** [comment]
Awesome, thank you for the update and cross-linking the issues
